src/index.js

```diff
@@ -37,11 +38,20 @@ import cartSVG from './static/cart.svg'
         creator.createButton();
     } else {
         const creator = new AppendCartButtonCreator(document, cartStorage, rimi);
-        const progressHandler = new CartUpdateProgressIndicator(document, rimi.refresh.bind(rimi));
-        creator.setProgressHandler(progressHandler);
+        const progressIndicator = new CartUpdateProgressIndicator(document, rimi.refresh.bind(rimi));
+        creator.setProgressHandler(progressIndicator);
         creator.createButtons(cartSVG);
     }
 
     const confirmer = new CartAbandonmentConfirmer(document, rimi.dom, promptService);
     confirmer.bindToCartChangeButtons();
+
+    let cartUpdate = cartStorage.popCartUpdate();
+    if (cartUpdate) {
+        let validator = new CartUpdateValidator(rimi.dom.getCurrentCart().products, cartUpdate);
+        if (validator.hasProductUpdateFailed()) {
+            let missingProductNames = validator.getMissingProductNames();
+            promptService.notifyProductAdditionFailed(missingProductNames);
```

Comment:
1. if multiple products are missing, the popup will look very poor
2. see if you can add links to the product pages (when a product addition failed, I would like to open the product page so that I can quickly locate an alternative product)